### ISSUE TITLE
URLs with %FX percentage encoding not forwarded via Port :8181 to ReGa but result in "400 Bad Request"

### ISSUE THREAD

**Issue by jens-maus** [jens-maus/RaspberryMatic#1831]
### Describe the issue you are experiencing

When performing the following curl command execution to write a text with "ö" umlaut in a system variable "TestRM" a `400 Bad Request` error is returned:

```html
# curl -vvv http://localhost:8181/hm.exe?value=dom.GetObject%28ID_SYSTEM_VARIABLES%29.Get%28%27TestRM%27%29.State%28%27Fl%F6te%27%29
> GET /hm.exe?value=dom.GetObject%28ID_SYSTEM_VARIABLES%29.Get%28%27TestRM%27%29.State%28%27Fl%F6te%27%29 HTTP/1.1
> Host: localhost:8181
> User-Agent: curl/7.79.1
> Accept: */*
> 
< HTTP/1.1 400 Bad Request
< Content-Type: text/html
< Content-Length: 345
< Connection: close
< Date: Mon, 09 May 2022 13:17:34 GMT
< 
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
         "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>400 Bad Request</title>
 </head>
 <body>
  <h1>400 Bad Request</h1>
 </body>
</html>
```

When removing the `%F6` in the above URL or replacing it with some other one, no Bad Request error is returned and the request will be handled correctly, thus forwarded to ReGaHss via its proxy statement.

### Describe the behavior you expected

The request should be internally forwarded to port 8183 which is the rega scripting port which lighttpd is proxying.

### Steps to reproduce the issue

1. login to CCU/RaspberryMatic via SSH
2. create a text system variable "TestRM"
3. execute `curl -vvv http://localhost:8181/hm.exe?value=dom.GetObject%28ID_SYSTEM_VARIABLES%29.Get%28%27TestRM%27%29.State%28%27Fl%F6te%27%29`
4. Monitor the output

### What is the version this bug report is based on?

3.63.9.20220430

### Which base platform are you running?

ova (Open Virtual Infrastructure)

### Which HomeMatic/homematicIP radio module are you using?

n/a

### Anything in the logs that might be useful for us?

The same issue can be seen when using any %FC in the URL.

Please note that any `%Fx` percentage URL encoding seems to trigger the issue. Thus no "%FC" or any other "%Fx" encoding with `x` corresponding to any hex value seems to be possible.

### Additional information

After some short investigation, the issue seems to be related to newer lighttpd versions normalizing their URLs to some extend and thus might end up in `400 Bad Request` for certain reserved characters. See here for more information/docs on that matter:

https://redmine.lighttpd.net/projects/lighttpd/wiki/Server_http-parseoptsDetails

After adding the `"url-normalize" => "disable"` statement the issue seems to be gone.

This also refs https://homematic-forum.de/forum/viewtopic.php?p=717789#p717791

**Comment by gstrauss** [comment]
Why iso-8859-1?  Why is it not converted to UTF-8 before url-encoding (%xx)?

lighttpd is checking that the URL is valid UTF-8.  More precisely, lighttpd quickly checks if the bytes are invalid UTF-8 bytes.

https://en.wikipedia.org/wiki/UTF-8#Comparison_with_other_encodings
> Fallback and auto-detection: Only a small subset of possible byte strings are a valid UTF-8 string: the bytes C0, C1, and F5 through FF cannot appear, and bytes with the high bit set must be in pairs, and other requirements.

https://en.wikipedia.org/wiki/UTF-8#Codepage_layout
> Do not appear in a valid UTF-8 sequence. C0 and C1 could be used only for an "overlong" encoding of a 1-byte character.[[22]](https://en.wikipedia.org/wiki/UTF-8#cite_note-26) F5 to FD are leading bytes of 4-byte or longer sequences that can only encode code points larger than U+10FFFF.[[23]](https://en.wikipedia.org/wiki/UTF-8#cite_note-27) FE and FF were never assigned any meaning.

[References: UTF-8, a transformation format of ISO 10646]
https://datatracker.ietf.org/doc/html/rfc3629#section-1
> The octet values C0, C1, F5 to FF never appear.

[References: Unicode Standard chapter 3]
https://www.unicode.org/versions/Unicode13.0.0/ch03.pdf
> As a consequence of the well-formedness conditions specified in Table 3-7, the following
byte values are disallowed in UTF-8: C0–C1, F5–FF.

**Comment by gstrauss** [comment]
> In previous versions of lighttpd this wasn't really an issue as it forwarded the %XX encoded URL portions verbatim without having touched them. But at some point it seems this is not the case anymore and now (as shown here) %FC or %F6 cannot be used anymore - which are however part of ISO-8859-1 (german umlauts).

To be clear, this _is not_ a limitation of UTF-8.  The entirety of ISO-8859-1 (single-byte encoding) and much more is able to be properly encoded in UTF-8 (multi-byte encoding).

Also, this _is not_ a bug in lighttpd.  The lighttpd behavior is explicit and intentional to support a stronger security stance.

The option for url normalization was added in lighttpd 1.4.50, released in Aug 2018.  After multiple announcements in release notes for subsequent releases, url normalization was enabled by default in lighttpd 1.4.54, released May 2019 (**3 years ago**).  

(&lt;generic rant&gt;) It is a frequent lament of mine and other open source developers that numerous popular linux distros are negligent (IMNSHO) in upgrading software in reasonable timeframes, and upgrading "stable" distros -- a.k.a. "nearly unmaintained" because "nearly unchanging" -- is not made easier for end-users, many of whom do not _manually_ perform _periodic_ upgrades on a _regular basis_ on their own.  There are many good reasons why browsers and other desktop software run tasks which check for and perform upgrades.  Similarly, automatic software upgrades (at least for security patches) is the default behavior for iOS and Android mobile devices. (&lt;/generic rant&gt;)

> > Why iso-8859-1? Why is it not converted to UTF-8 before url-encoding (%xx)?

> Well, the application which finally receives the HTTP request (behind the lighttpd proxy) is a very old-school (199x) application that can only really speak iso-8859-1 and thus expects also URLs to be percentage-encoded in with ISO-8859-1 and not UTF-8.

What is producing the URL?  What is percent-encoding the problematic ISO-8859-1 string?  Is it the ancient app or is it something else?  Whatever it is that is doing the percent-encoding, it might be appropriate to modify that code to convert the string to UTF-8 prior to percent-encoding.  Then, you could use lighttpd mod_magnet for requests only to this application, percent-decode the query-string, convert to ISO-8859-1, and re-percent-encode before passing the request on to the application.  Alternatively, modify the application to base64-encode/decode the iso-8859-1 string into the URL query-string instead of selective percent-encoding.  Alternatively, use POST to send the data instead of using query-string.  Alternatively, use an HTTP header instead of using the query-string, and use `"header-strict" => "disable"` while leaving `"url-normalize" => "enable"`.

Best practices strongly recommend UTF-8 in URIs since there is no generic way to convey the charset used in the URI, whereas there are alternatives to specify charset for strings in HTTP headers.  Search for "UTF-8" in https://html.spec.whatwg.org/ and note the frequency of the explicit phrase "UTF-8 percent-encode" https://url.spec.whatwg.org/#string-utf-8-percent-encode

> Nevertheless setting `"url-normalize" => "disable"` in the lighttpd.conf solves this issue currently. Question remains, however: Is this really the way to go and which potential side-effects this could have?

No, it is not "the way to go" for a security-focused solution.  A security-focused solution would address the source of the issue: a 199x application that does not support UTF-8.

Instead, it is a "quick-fix" **workaround** to use `"url-normalize" => "disable"` to disable lighttpd URL normalization (which includes detection of invalid UTF-8 in percent-encodings).

> which potential side-effects this could have?

The lighttpd core provides the `"url-normalize"` option and works with or without that feature enabled.  However, depending on your lighttpd.conf configuration, you might not get the behavior you expect in some situations.  Your mod_rewrite and mod_redirect rules might not match for _non-normalized_ alternative encodings.  Best practices suggest writing explicit allow rules, and then deny everything else, but I am sure there are many custom lighttpd.conf instances that do otherwise.

Here is one example of unwanted behavior that is fixed with url normalization: https://redmine.lighttpd.net/issues/1720

---

tl;dr: if fixing the application is not an option, then using `"url-normalize" => "disable"` is a **workaround**, though disabling the url normalization may cause lighttpd.conf configured behavior to change for _non-normalized_ URIs, which in turn may have security implications for your specific environment.

If fixing the application is an option, who would do that and where is the code?  Here is an example for charset conversion for Python and Java: https://mincong.io/2019/04/07/understanding-iso-8859-1-and-utf-8/

Given the **3 years** that has elapsed since lighttpd enabled url-normalization by default, is it worthwhile for me to even consider any enhancements to lighttpd url-normalization options, which may take another 3-4 years to reach end-users?

**Comment by gstrauss** [comment]
Is hm.exe open-source?  Where is the source code for hm.exe?  What package provides it?

**Comment by jens-maus** [comment]
> Is hm.exe open-source? Where is the source code for hm.exe? What package provides it?

`hm.exe` is just an arbitrary string in the query URL with a `*.exe` extension sent over to port 8181. Once this HTTP request reaches the internal webserver of the closed source `ReGaHss` application, this application interprets the query string value as a special scripting call. In fact, while ReGaHss is listening itself to port 8183 (only on localhost), lighttpd is used as a transparent proxy also providing some kind of authentication and listens on port 8181 (cf. https://github.com/jens-maus/RaspberryMatic/blob/master/buildroot-external/overlay/base/etc/lighttpd/conf.d/proxy_normal.conf#L84-L87).

So sorry, but this `ReGaHss` application is indeed unfortunately closed source. Nevertheless, it heavily relies on the query URL being URL percentage encoded with the `%xx` percentage encoding staying untouched by lighttpd, so that each `%xx` is then converted into its non-hexadecimal counterpart within the URL conversion routines within `ReGaHss` itself. And this application is unfortunately only ISO-8859-1 (latin1) aware, thus expects the URL percentage encoding to be latin1 encoded as well.

Hope that clarifies it.

**Comment by gstrauss** [comment]
I notice in https://github.com/jens-maus/RaspberryMatic/blob/master/buildroot-external/overlay/base/etc/lighttpd/conf.d/mime.conf (last updated about 3 years ago) that the charset is not specified with the text mime-types, e.g. "text/html;charset=utf-8".  Recent versions of lighttpd specify charset=utf-8 by default with the mime-type for text mime-types for lighttpd-provided `doc/config/conf.d/mime.conf` or lighttpd-provided `doc/scripts/create-mime.conf.pl`.  ReGaHss might be configured to send "Content-Type: text/html;charset=iso-8859-1", or to specify the charset in the HTML.  I wonder (**untested**) whether or not many/some/all/none client browsers would rewrite links to UTF-8 before following those links.

**Comment by jens-maus** [comment]
> Thank you for that extra info. If it were something I had access to, it would probably only be a small amount of effort to translate from UTF-8 to ISO-8859-1.

Well, in principle I could do this (I do have access to that closed source app). However, as said before: This is a very old application which provides this API to a various of outside applications which exhibit this ISO-8859-1 URL encoding possibility already. Changing this now would probably require all these applications to be changed as well and would render the API partly incompatible, I guess. Or I would have to implement some automatic identification of the used encoding in the URL (which I actually don't know really how to reliably do).

BTW: Where exactly is it written that URL encoding should be UTF-8 encoded? Until now, I always thought URL percentage encoding is only a vebatim hexadecimal character encoding regardless of any charset information, thus simply encoding/decoding %xx to/from its hexadecimal interpretation.

> Right now, it is straightforward in lighttpd mod_magnet to write a lua script which takes a UTF-8 percent-encoded query string and translates it to an ISO-8859-1 percent-encoded query string before passing the request to the backend.

This sound interested, but RaspberryMatic (in which ReGaHss+lighttpd is integrated) is using a buildroot/Linux environment for embedded platforms, thus does not come with lua scripting support (yet).

> Is ReGaHss the application that is producing the links with the ISO-8859-1 percent-encoded query strings?

No, ReGaHss is "just" provding the web service which access these "*.exe" URLs which can contain these percentage coded URL parameters, but which must be ISO-8859-1 encoded, thought. As said before, ReGaHss is just providing the API in terms of this web service and there are tons of users out there using this API relying on its current behaviour (1:1 hexadeciaml percentage URL encoding/decoding).

> What (approximate) sizes are the responses from ReGaHss? Something that can fit into memory? Or too large to fit into memory? In lighttpd 1.4.65, it will be possible to capture the backend response in lighttpd mod_magnet and to then to have a lua script rewrite links, though that would take more effort.

The response ReGaHss will give back will certainly fit into memory. Nothing too large really.

> I notice in https://github.com/jens-maus/RaspberryMatic/blob/master/buildroot-external/overlay/base/etc/lighttpd/conf.d/mime.conf (last updated about 3 years ago) that the charset is not specified with the text mime-types, e.g. "text/html;charset=utf-8". Recent versions of lighttpd specify charset=utf-8 by default with the mime-type for text mime-types for lighttpd-provided `doc/config/conf.d/mime.conf` or lighttpd-provided `doc/scripts/create-mime.conf.pl`.

Sorry, don't get it really. I can see that the `charset=XXX` is missing in `mime.conf`. But this shouldn't have any impact. Or did I miss something?

> ReGaHss might be configured to send "Content-Type: text/html;charset=iso-8859-1", or to specify the charset in the HTML. I wonder (**untested**) whether or not many/some/all/none client browsers would rewrite links to UTF-8 before following those links.

As said before, ReGaHss does only interpret iso-8859-1 and will also provide all its HTTP responses in iso-8859-1 encoded form, also with sending its response using `Content-type: text/html; charset=iso-8859-1` in the response header.

**Comment by jens-maus** [comment]
> To be clear, my goal in continuing this conversation is to see if there is a low-effort workaround which might be available with a little effort of mine, and possibly better than disabling lighttpd's url-normalization.

I am totally with you here! I fully understand that proper URL normalization also helps in mitigating against certain attack vectors like e.g. sending over %00 encoded strings in the hope to provoke a buffer overflow or similar.

> lighttpd default url-normalization is too strict for ReGaHss, and lighttpd provides a config option to disable lighttpd's url-normalization. I too a (very) quick look through RasberryMatic lighttpd configs and did not immediately see use of mod_rewrite or mod_redirect which might be affected by disabling lighttpd url-normalization.

Thanks for having checked. If there would be a possibility to only disable the URL normalization for a certain `$SERVER["socket"]` part in the lighttpd.conf this could already help, indeed. If possible I could then only disable the normalization within https://github.com/jens-maus/RaspberryMatic/blob/master/buildroot-external/overlay/base/etc/lighttpd/conf.d/proxy_normal.conf#L84-L87 and omit disabling it for the rest (port 80, port 443) which would offer some security benefits and could be a better compromise than globally disabling url normalization in lighttpd.

**Comment by gstrauss** [comment]
Unfortunately, there is a chicken-and-egg problem in using the non-normalized URL to configure lighttpd on whether or not to normalize the URL.  With $SERVER["socket"], that is clear, but it is not clear with other lighttpd conditionals.  That is why the url-normalization is a global setting, as it is applied to the request headers before the request headers are used to evaluate lighttpd conditional configuration.  (Yes, it is true that the socket is known beforehand, but lighttpd currently avoids multiple evaluations of the conditional configuration.)

Instead, I might be able to provide a url-normalization configuration option to normalize the url-path, but not the query-string.  Still, I am concerned that it would take **years** to reach end-users due to the extremely slow propagation of updates from Debian and Ubuntu through RPi.

**Comment by jens-maus** [comment]
> Instead, I might be able to provide a url-normalization configuration option to normalize the url-path, but not the query-string. Still, I am concerned that it would take **years** to reach end-users due to the extremely slow propagation of updates from Debian and Ubuntu through RPi.

Well, for this application here (RaspberryMatic) this is not an issue at all, since I can fully decide myself which lighttpd version I am going to integrate with the next RaspberryMatic version (released monthly) and I have also already integrated newer lighttpd versions as fixes were released for things I was interested in. That's why I do have a `buildroot-patches` top-level directory in which I maintain upstream patches against the used buildroot version (cf. https://github.com/jens-maus/RaspberryMatic/tree/master/buildroot-patches).

So if would release such an improvement I would be happy to use it in the near future (and would provide feedback on it, of course).

**Comment by gstrauss** [comment]
> Well, for this application here (RaspberryMatic) this is not an issue at all, since I can fully decide myself which lighttpd version I am going to integrate with the next RaspberryMatic version (released monthly) and I have also already integrated newer lighttpd versions as fixes were released for things I was interested in.

Ok.  I'll mock up some patches in the next couple weeks.  I think I can do so without adversely affecting lighttpd performance (url-normalization does have a non-zero cost, but normalization is worth it -- though the lighttpd executable may grow 1-2k.  Would you like me to post here, or open a new issue, or should I post (here) a link to a topic on the lighttpd forum?

**Comment by jens-maus** [comment]
> Ok. I'll mock up some patches in the next couple weeks. I think I can do so without adversely affecting lighttpd performance -- url-normalization does have a non-zero cost, but normalization is worth it -- though the lighttpd executable may grow 1-2k. Would you like me to post here, or open a new issue, or should I post (here) a link to a topic on the lighttpd forum?

This sounds great. Do whatever fits you best regarding management of that request. But at least putting here a reference to your issue or commit would be great, indeed. So thanks in advance for all your work and for taking care of lighttpd at all. Great, slim web server it is, btw. :)

**Comment by gstrauss** [comment]
FYI: `buildroot-external/package/tclrega/tclrega.cpp` appears to be aware of and convert between iso-8859-1 and UTF-8

**Comment by gstrauss** [comment]
On my **development** branch, where things might change, the top commit may be cherry-picked and applied to lighttpd-1.4.64.
https://git.lighttpd.net/lighttpd/lighttpd1.4/src/branch/personal/gstrauss/master
https://git.lighttpd.net/lighttpd/lighttpd1.4/commit/f529566baf5e8c329d7b3f5bd52532e5aa07cd67
The default is `server.http-parseopts += ("url-invalid-utf8-reject" => "enable")` to preserve existing behavior.
